File: server/verify_parity.py
"""verify_parity.py - prove the simulated board matches the real game.

For a set of (difficulty, seed, first-click) triples this tool:
  1. starts minesweeper-x64.exe --listen <port>,
  2. applies the seed as a persistent Normal seed and plays the first click
     through the scripting CLI,
  3. builds the same board in sim_engine and clicks the same cell,
  4. compares the two `board` dumps cell-by-cell (and `opened`/`over`).

A mismatch means the sim's xorshift64 / place_mines / reveal logic drifted
from the C code.  Run from the server/ directory (Windows only; needs the
compiled exe).

Usage:
  python verify_parity.py [--exe ..\\build\\minesweeper-x64.exe]
      [--port 31350] [--seeds 8] [--fast]
"""

import argparse
import os
import random
import socket
import subprocess
import sys
import time

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
sys.path.insert(0, HERE)

from sim_engine import SimBoard  # noqa: E402

DIFFS = ["beginner", "intermediate", "expert"]
# (rows, cols) per difficulty, used to pick first-click positions.
SIZES = {"beginner": (8, 8), "intermediate": (16, 16), "expert": (16, 30)}
FIRST_CLICKS = ["center", "corner", "edge"]


def first_cell(diff, pos):
    rows, cols = SIZES[diff]
    if pos == "center":
        return rows // 2, cols // 2
    if pos == "corner":
        return 0, 0
    return rows // 2, 0


def find_free_port(base):
    for p in range(base, base + 200):
        s = socket.socket()
        try:
            s.bind(("127.0.0.1", p))
            s.close()
            return p
        except OSError:
            s.close()
    raise RuntimeError("no free port near %d" % base)


def wait_port(port, exe_proc, timeout=15):
    t0 = time.time()
    while time.time() - t0 < timeout:
        if exe_proc.poll() is not None:
            raise RuntimeError("exe exited early with code %r" %
                               exe_proc.returncode)
        try:
            s = socket.create_connection(("127.0.0.1", port), timeout=1)
            s.close()
            return
        except OSError:
            time.sleep(0.1)
    raise RuntimeError("game did not open port %d" % port)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--exe", default=os.path.join(ROOT, "build",
                                                  "minesweeper-x64.exe"))
    ap.add_argument("--port", type=int, default=0)
    ap.add_argument("--seeds", type=int, default=8)
    ap.add_argument("--fast", action="store_true",
                    help="skip expert/edge combos for a quicker pass")
    args = ap.parse_args()

    sys.path.insert(0, os.path.join(ROOT, "minesweeper_bot"))
    from ms_client import MSClient

    if not os.path.exists(args.exe):
        print("exe not found: %s (build it first)" % args.exe)
        return 2
    port = args.port or find_free_port(31350)

    exe = subprocess.Popen([args.exe, "--listen", str(port)],
                           stdout=subprocess.DEVNULL,
                           stderr=subprocess.DEVNULL)
    try:
        wait_port(port, exe)
        client = MSClient(port)
        if not client.ping():
            print("game did not answer ping")
            return 2

        rng = random.Random(20260209)
        fails = 0
        total = 0
        combos = []
        for diff in DIFFS:
            positions = FIRST_CLICKS[:]
            if args.fast and diff == "expert":
                positions = ["center"]
            for pos in positions:
                for _ in range(args.seeds):
                    combos.append((diff, pos, rng.randrange(0, 1 << 63)))

        for diff, pos, seed in combos:
            total += 1
            r0, c0 = first_cell(diff, pos)
            client.seed_diff(diff, seed)
            client.new(diff)
            client.click(r0, c0)
            live_state = client.state()
            live_board = client.board()

            sim = SimBoard()
            sim.new(diff, seed)
            sim.click(r0, c0)
            sim_board = sim.board()
            sim_state = sim.state()

            mismatches = []
            if sim_board != live_board:
                mismatches.append("board differs")
            for key in ("opened", "over", "started"):
                if sim_state.get(key) != live_state.get(key):
                    mismatches.append("%s %s!=%s" % (
                        key, sim_state.get(key), live_state.get(key)))
            if mismatches:
                fails += 1
                print("  MISMATCH %s seed=%d first=(%d,%d) pos=%s: %s"
                      % (diff, seed, r0, c0, pos, "; ".join(mismatches)))
                for i, (a, b) in enumerate(zip(live_board, sim_board)):
                    if a != b:
                        print("    live row %2d: %s" % (i, a))
                        print("    sim  row %2d: %s" % (i, b))
                        break

        print("parity: %d/%d boards identical (%d mismatch%s)"
              % (total - fails, total, fails, "es" if fails != 1 else ""))
        client.close()
        return 0 if fails == 0 else 1
    finally:
        try:
            exe.terminate()
        except OSError:
            pass


if __name__ == "__main__":
    sys.exit(main())

File: ms/cli/verify-parity.js
// verify-parity.js - prove the simulated board matches the real game.
//
// Zero-dependency port of server/verify_parity.py.  For a set of
// (difficulty, seed, first-click) triples this tool:
//   1. starts minesweeper-x64.exe --listen <port>,
//   2. applies the seed as a persistent Normal seed and plays the first click
//      through the scripting CLI,
//   3. builds the same board in SimBoard and clicks the same cell,
//   4. compares the two `board` dumps cell-by-cell (and `opened`/`over`).
//
// A mismatch means the sim's xorshift64 / place_mines / reveal logic drifted
// from the C code.  Run from the ms/ directory or repo root (Windows only;
// needs the compiled exe).

import { spawn } from "node:child_process";
import net from "node:net";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";
import { Random } from "../core/mt19937.js";
import { SimBoard } from "../core/sim-engine.js";
import { MSClient } from "../core/client.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const REPO = path.resolve(HERE, "..", "..");
const DEFAULT_EXE = path.join(REPO, "build", "minesweeper-x64.exe");

const DIFFS = ["beginner", "intermediate", "expert"];
// (rows, cols) per difficulty, used to pick first-click positions.
const SIZES = { beginner: [8, 8], intermediate: [16, 16], expert: [16, 30] };
const FIRST_CLICKS = ["center", "corner", "edge"];

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function firstCell(diff, pos) {
  const [rows, cols] = SIZES[diff];
  if (pos === "center") return [Math.floor(rows / 2), Math.floor(cols / 2)];
  if (pos === "corner") return [0, 0];
  return [Math.floor(rows / 2), 0];
}

function findFreePort(base) {
  return new Promise((resolve, reject) => {
    const tryBind = (p) => {
      const srv = net.createServer();
      srv.once("error", () => {
        if (p < base + 200) tryBind(p + 1);
        else reject(new Error("no free port near " + base));
      });
      srv.listen(p, "127.0.0.1", () => {
        const port = srv.address().port;
        srv.close(() => resolve(port));
      });
    };
    tryBind(base);
  });
}

async function waitPort(port, exeProc, timeout = 15000) {
  const t0 = Date.now();
  while (Date.now() - t0 < timeout) {
    if (exeProc.exitCode !== null) {
      throw new Error("exe exited early with code " + exeProc.exitCode);
    }
    const ok = await new Promise((resolve) => {
      const s = net.connect(port, "127.0.0.1");
      s.once("connect", () => {
        s.destroy();
        resolve(true);
      });
      s.once("error", () => resolve(false));
      s.setTimeout(1000, () => {
        s.destroy();
        resolve(false);
      });
    });
    if (ok) return;
    await sleep(100);
  }
  throw new Error("game did not open port " + port);
}

export async function run(args) {
  if (!args.exe || !exists(args.exe)) {
    console.log("exe not found: " + args.exe + " (build it first)");
    return 2;
  }
  const port = args.port || (await findFreePort(31350));

  const exe = spawn(args.exe, ["--listen", String(port)], {
    stdio: "ignore",
  });
  try {
    await waitPort(port, exe);
    const client = new MSClient(port);
    if (!(await client.ping())) {
      console.log("game did not answer ping");
      return 2;
    }

    const rng = new Random(20260209);
    let fails = 0;
    let total = 0;
    const combos = [];
    for (const diff of DIFFS) {
      const positions = FIRST_CLICKS.slice();
      if (args.fast && diff === "expert") positions.length = 1; // center only
      for (const pos of positions) {
        for (let i = 0; i < args.seeds; i++) {
          combos.push([diff, pos, rng.randrange(0n, 1n << 63n)]);
        }
      }
    }

    for (const [diff, pos, seed] of combos) {
      total += 1;
      const [r0, c0] = firstCell(diff, pos);
      const attempt = async () => {
        await client.seedDiff(diff, seed);
        await client.new(diff);
        await client.click(r0, c0);
        const liveState = await client.state();
        const liveBoard = await client.board();

        const sim = new SimBoard();
        sim.new(diff, seed);
        sim.click(r0, c0);
        const simBoard = sim.board();
        const simState = sim.state();

        const mismatches = [];
        if (simBoard.join("\n") !== liveBoard.join("\n")) mismatches.push("board differs");
        for (const key of ["opened", "over", "started"]) {
          if (simState[key] !== liveState[key]) {
            mismatches.push(`${key} ${simState[key]}!=${liveState[key]}`);
          }
        }
        return { mismatches, liveBoard, simBoard };
      };

      let result = await attempt();
      if (result.mismatches.length) {
        // The live game occasionally drops a scripting click (opened=0,
        // started=0) while the sim proceeded; re-verify the same combo once
        // before treating a difference as a real parity failure.
        result = await attempt();
      }
      if (result.mismatches.length) {
        fails += 1;
        console.log(
          `  MISMATCH ${diff} seed=${seed} first=(${r0},${c0}) pos=${pos}: ${result.mismatches.join("; ")}`,
        );
        for (let i = 0; i < result.liveBoard.length && i < result.simBoard.length; i++) {
          if (result.liveBoard[i] !== result.simBoard[i]) {
            console.log(`    live row ${i}: ${result.liveBoard[i]}`);
            console.log(`    sim  row ${i}: ${result.simBoard[i]}`);
            break;
          }
        }
      }
    }

    console.log(
      `parity: ${total - fails}/${total} boards identical (${fails} mismatch${fails === 1 ? "" : "es"})`,
    );
    await client.close();
    return fails === 0 ? 0 : 1;
  } finally {
    try {
      exe.kill();
    } catch {
      // ignore
    }
  }
}

function exists(p) {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
}

function usage() {
  console.log(
    "usage: node ms/cli/verify-parity.js [--exe ..\\build\\minesweeper-x64.exe] " +
      "[--port 31350] [--seeds 8] [--fast]",
  );
}

async function main() {
  const { values, positionals } = parseArgs({
    options: {
      exe: { type: "string", default: DEFAULT_EXE },
      port: { type: "string", default: "0" },
      seeds: { type: "string", default: "8" },
      fast: { type: "boolean", default: false },
      help: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });

  if (values.help || positionals.length) {
    usage();
    return values.help ? 0 : 2;
  }

  return run({
    exe: values.exe,
    port: Number(values.port),
    seeds: Number(values.seeds),
    fast: values.fast,
  });
}

const isMain =
  process.argv[1] !== undefined &&
  path.resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (isMain) {
  process.exit(await main());
}

export { main };
